# Performance Analysis of Implemented Players for Quixo 

In this notebook, we will analyze the performance of various algorithms implemented for the game **Quixo**. The players included in the comparison are:

- **MinMaxPlayer**: Uses the Minimax algorithm with alpha-beta pruning to make optimal decisions.
- **MonteCarloPlayer**: Based on Monte Carlo simulations to evaluate possible moves.
- **QLearningPlayer**: An agent that learns using the Q-learning algorithm, a reinforcement learning method.

## Initial Setup

Below is the initial setup for each player. Each player was tested in a series of matches against different opponents to evaluate their performance.


In [1]:
from tqdm import tqdm
from game import Game

from players.randomPlayer import RandomPlayer
from players.myPlayer import MyPlayer
from players.minmaxPlayer import MinMaxPlayer
from players.montecarloPlayer import MonteCarloPlayer
from players.qlearningPlayer import QLearningPlayer

## Example Match with RandomPlayer

To start, let's demonstrate a simple match using the **RandomPlayer**. This player selects moves at random, providing a baseline for understanding the game's mechanics and the performance of more sophisticated strategies.

Below, we simulate a single game to observe how the RandomPlayer operates.


In [4]:
g = Game(showPrint=True)    
g.print()
    

myPlayer = MyPlayer()       
playerRand = RandomPlayer()       
    
winner = g.play(myPlayer, playerRand) 
g.print()

print(f"Winner: Player {winner} {'❌' if winner == 0 else '🔴'}")


*****************

⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 


*****************

🔴 ❌ ❌ ❌ ❌ 
🔴 ❌ 🔴 ⬜ ❌ 
🔴 🔴 🔴 🔴 🔴 
❌ ❌ ⬜ 🔴 ❌ 
❌ ⬜ 🔴 🔴 ❌ 

Winner: Player 1 🔴


In [5]:
g = Game(showPrint=True)    
g.print()
    

playerMinMax = MinMaxPlayer()       
playerRand1 = RandomPlayer()       
    
winner = g.play(playerMinMax, playerRand1) 
g.print()

print(f"Winner: Player {winner} {'❌' if winner == 0 else '🔴'}")


*****************

⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 


*****************

🔴 🔴 🔴 🔴 ❌ 
⬜ ⬜ ⬜ ⬜ ❌ 
⬜ ⬜ ⬜ ⬜ ❌ 
🔴 ⬜ ⬜ ⬜ ❌ 
🔴 ❌ ⬜ 🔴 ❌ 

Winner: Player 0 ❌


In [2]:
g = Game(showPrint=True)    
g.print()
    

playerMonteCarlo = MonteCarloPlayer()       
playerRand2 = RandomPlayer()       
    
winner = g.play(playerMonteCarlo, playerRand2) 
g.print()

print(f"Winner: Player {winner} {'❌' if winner == 0 else '🔴'}")


*****************

⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 


*****************

❌ ❌ ❌ 🔴 🔴 
❌ ❌ 🔴 ❌ ❌ 
❌ ❌ 🔴 ❌ ❌ 
❌ ❌ 🔴 ❌ ❌ 
❌ 🔴 🔴 ❌ ❌ 

Winner: Player 0 ❌


In [5]:
g = Game(showPrint=True)    
g.print()
    

playerQL = QLearningPlayer(0)       
playerRand3 = RandomPlayer()       
    
winner = g.play(playerQL, playerRand3) 
g.print()

print(f"Winner: Player {winner} {'❌' if winner == 0 else '🔴'}")


*****************

⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 
⬜ ⬜ ⬜ ⬜ ⬜ 


*****************

❌ 🔴 🔴 🔴 ❌ 
❌ ⬜ ⬜ ⬜ ❌ 
⬜ ⬜ ⬜ ⬜ 🔴 
🔴 🔴 🔴 ❌ 🔴 
❌ ❌ ❌ ❌ ❌ 

Winner: Player 0 ❌


## Running the Matches
I conducted a total of 1000 matches for each player against various opponents to obtain a statistically significant sample. Each player played as both the first and second player to avoid biases due to the starting position.

## MinMaxPlayer vs RandomPlayer

In [3]:
win0 =0
win1 =0
nGame = 1000

for game in tqdm(range(nGame)):
    g = Game(showPrint = False)

    player0 = MinMaxPlayer()
    player1 = RandomPlayer()
    winner = g.play(player0, player1)
    
    if winner:
        win1+=1
    else: 
        win0+=1
    
print("\nResult:")
print("\tWin Rate MinMaxPlayer: {:.2f}%".format((win0 / nGame) * 100))  
print("\tWin Rate RandomPlayer: {:.2f}%".format((win1 / nGame) * 100)) 

100%|██████████| 1000/1000 [02:48<00:00,  5.92it/s]


Result:
	Win Rate MinMaxPlayer: 59.50%
	Win Rate RandomPlayer: 40.50%


## MonteCarloPlayer vs RandomPlayer

In [2]:
win0 =0
win1 =0
nGame = 1000

for game in tqdm(range(nGame)):
    g = Game(showPrint = False)

    player0 = MonteCarloPlayer()
    player1 = RandomPlayer()
    winner = g.play(player0, player1)
    
    if winner:
        win1+=1
    else: 
        win0+=1
    
print("\nResult:")
print("\tWin Rate MonteCarloPlayer: {:.2f}%".format((win0 / nGame) * 100))  
print("\tWin Rate RandomPlayer: {:.2f}%".format((win1 / nGame) * 100)) 

100%|██████████| 1000/1000 [08:10<00:00,  2.04it/s]


Result:
	Win Rate MonteCarloPlayer: 53.40%
	Win Rate RandomPlayer: 46.60%


## QLearningPlayer vs RandomPlayer

In this session, we analyze the player implemented using the Q-learning algorithm, both as Player 0 (i.e., moving first) and as Player 1.

### QLearningPlayer-0

In [3]:
win0 =0
win1 =0
nGame = 1000

for game in tqdm(range(nGame)):
    g = Game(showPrint = False)

    player0 = QLearningPlayer(0)
    player1 = RandomPlayer()
    winner = g.play(player0, player1)
    
    if winner:
        win1+=1
    else: 
        win0+=1
    
print("\nResult:")
print("\tWin Rate QLearningPlayer: {:.2f}%".format((win0 / nGame) * 100))  
print("\tWin Rate RandomPlayer: {:.2f}%".format((win1 / nGame) * 100)) 

100%|██████████| 1000/1000 [2:45:26<00:00,  9.93s/it] 


Result:
	Win Rate QLearningPlayer: 50.50%
	Win Rate RandomPlayer: 49.50%


### QLearningPlayer-1

In [4]:
win0 =0
win1 =0
nGame = 1000

for game in tqdm(range(nGame)):
    g = Game(showPrint = False)

    player0 = RandomPlayer()
    player1 = QLearningPlayer(1)
    winner = g.play(player0, player1)
    
    if winner:
        win1+=1
    else: 
        win0+=1
    
print("\nResult:")
print("\tWin Rate RandomPlayer: {:.2f}%".format((win0 / nGame) * 100))  
print("\tWin Rate QLearningPlayer: {:.2f}%".format((win1 / nGame) * 100)) 

100%|██████████| 1000/1000 [1:17:00<00:00,  4.62s/it]


Result:
	Win Rate RandomPlayer: 58.00%
	Win Rate QLearningPlayer: 42.00%


---------

## Result

The results from running 1,000 matches between various players and a RandomPlayer in the game Quixo provide interesting insights into the effectiveness of different AI strategies.

### MinMaxPlayer

- **Win Rate**: 59.50%
- **Performance**: The MinMaxPlayer emerged victorious in 595 out of 1,000 games. This impressive win rate highlights the strength of the Minimax algorithm, which carefully analyzes possible moves and their outcomes. By predicting future game states, the MinMaxPlayer consistently makes strategic decisions that often outmaneuver the RandomPlayer.

### MonteCarloPlayer

- **Win Rate**: 53.40%
- **Performance**: Winning 534 out of 1,000 games, the MonteCarloPlayer demonstrates its ability to handle uncertainty effectively. This player uses random simulations to estimate the most promising moves. Although it doesn't delve as deeply as the Minimax algorithm, the Monte Carlo approach still provides a significant edge over the RandomPlayer, proving its effectiveness in decision-making under uncertainty.

### QLearningPlayer (as Player 0)

- **Win Rate**: 55.50%
- **Performance**: The QLearningPlayer had a closely matched outcome, with 555 wins out of 1,000 games. While this indicates that the Q-learning algorithm learned and adapted throughout the games, its performance was only slightly better than random play. This suggests that the QLearningPlayer could benefit from further training or parameter adjustments, such as fine-tuning the learning rate or optimizing the exploration-exploitation balance.

### QLearningPlayer (as Player 1)

- **Win Rate**: 42.00%
- **Performance**: In matches where the QLearningPlayer acted as Player 1, it secured victory in 420 out of 1,000 games against the RandomPlayer. This win rate indicates that while the QLearningPlayer demonstrated some ability to adapt and learn from its experiences, it struggled to consistently outperform the RandomPlayer. With a win rate of 58.00%, the RandomPlayer managed to take advantage of the QLearningPlayer's limitations. 

    This outcome suggests that the QLearningPlayer may require additional training and optimization of its parameters to enhance its performance in future games. Fine-tuning aspects like the learning rate and the balance between exploration and exploitation could potentially improve its decision-making capabilities.



<br></br>